tokio-quiche/src/quic/addr_validation_token.rs

```diff
@@ -54,140 +51,135 @@ impl Default for AddrValidationTokenManager {
 
 impl AddrValidationTokenManager {
     pub(super) fn gen(
-        &self, original_dcid: &[u8], client_addr: SocketAddr,
+        &self, original_dcid: &[u8], client_addr: SocketAddr, scid: &[u8],
     ) -> Vec<u8> {
         let ip_bytes = match client_addr.ip() {
-            IpAddr::V4(addr) => addr.octets().to_vec(),
-            IpAddr::V6(addr) => addr.octets().to_vec(),
+            IpAddr::V4(addr) => &addr.octets()[..],
+            IpAddr::V6(addr) => &addr.octets()[..],
         };
 
-        let token_len = HMAC_TAG_LEN + ip_bytes.len() + original_dcid.len();
-        let mut token = io::Cursor::new(vec![0u8; token_len]);
-
-        token.set_position(HMAC_TAG_LEN as u64);
-        token.write_all(&ip_bytes).unwrap();
-        token.write_all(original_dcid).unwrap();
+        // The token itself only consists of the client's original DCID and the
+        // HMAC tag. However, we calculate the HMAC over the client's IP and our
+        // SCID as well to bind the token to a specific connection.
+        let hmac_len = std::cmp::max(
+            original_dcid.len() + HMAC_TAG_LEN,
+            original_dcid.len() + ip_bytes.len() + scid.len(),
+        );
 
-        let tag = boring::hash::hmac_sha256(
-            &self.sign_key,
-            &token.get_ref()[HMAC_TAG_LEN..],
-        )
-        .unwrap();
+        let mut token_buf = Vec::with_capacity(hmac_len);
+        token_buf.extend_from_slice(original_dcid);
+        token_buf.extend_from_slice(ip_bytes);
+        token_buf.extend_from_slice(scid);
 
-        token.set_position(0);
-        token.write_all(tag.as_ref()).unwrap();
+        let tag = boring::hash::hmac_sha256(&self.sign_key, &token_buf).unwrap();
+        debug_assert_eq!(tag.len(), HMAC_TAG_LEN);
 
-        token.into_inner()
+        // Drop the non-payload parts of the HMAC and reuse the storage for the
```

Comment:
If we are dropping the non-CID parts of the buffer anyway I think we can just use the `Hasher::update()` API instead to avoid having to allocate the whole `token_buf` and then truncating it https://docs.rs/boring/latest/boring/hash/struct.Hasher.html#method.update